xen/arch/arm/coproc/coproc.h

```diff
@@ -171,52 +271,53 @@ enum COPROC_DBG_LEVEL
     coproc_dev_print(dev, XENLOG_INFO,    COPROC_DBG_INFO, fmt, ## __VA_ARGS__)
 #define COPROC_DEBUG(dev, fmt, ...)                                            \
     coproc_dev_print(dev, XENLOG_DEBUG,   COPROC_DBG_DEBUG, fmt, ## __VA_ARGS__)
-#define COPROC_VERBOSE(dev, fmt, ...)                                            \
+#define COPROC_VERBOSE(dev, fmt, ...)                                          \
     coproc_dev_print(dev, XENLOG_DEBUG,   COPROC_DBG_VERB, fmt, ## __VA_ARGS__)
 
 void coproc_init(void);
-struct coproc_device * coproc_alloc(struct dt_device_node *,
-                                    const struct coproc_ops *);
-int coproc_register(struct coproc_device *);
-void coproc_release(struct coproc_device *);
-struct vcoproc_instance *coproc_get_vcoproc(struct domain *,
-                                            struct coproc_device *);
+struct mcoproc_device * coproc_alloc(struct dt_device_node *,
+                                     const struct pcoproc_desc *,
+                                     const struct coproc_ops *);
+int coproc_register(struct mcoproc_device *);
+void coproc_release(struct mcoproc_device *);
 int vcoproc_domain_init(struct domain *);
 void vcoproc_domain_free(struct domain *);
 int coproc_do_domctl(struct xen_domctl *, struct domain *,
                      XEN_GUEST_HANDLE_PARAM(xen_domctl_t));
-bool_t coproc_is_attached_to_domain(struct domain *, const char *);
 void vcoproc_continue_running(struct vcoproc_instance *);
 int coproc_release_vcoprocs(struct domain *);
+int vcoproc_handle_node(struct domain *, void *, const struct dt_device_node *);
 
 #define dev_path(dev) dt_node_full_name(dev_to_dt(dev))
 
-static inline void vcoproc_get_rw_context(struct domain *d, struct mmio *mmio,
-                                          mmio_info_t *info,
-                                          struct vcoproc_rw_context *ctx)
+static inline void
+vcoproc_get_rw_context(struct domain *d, struct vcoproc_mmio *v_mmio,
+                       mmio_info_t *info, struct vcoproc_rw_context *ctx)
 {
-    ctx->coproc = mmio->coproc;
     ctx->dabt = info->dabt;
-    ctx->offset = info->gpa - mmio->addr;
-    ctx->vcoproc = coproc_get_vcoproc(d, ctx->coproc);
+    ctx->offset = info->gpa - v_mmio->addr;
+    ctx->vcoproc = v_mmio->vcoproc;
     BUG_ON(ctx->vcoproc == NULL);
```

Comment:
Don't think you need to keep BUG_ON anymore. @andr2000 had added it because coproc_get_vcoproc() may return NULL.